### ISSUE TITLE
[ENH] Add Spline Forecaster

### ISSUE THREAD

**Issue by Garve** [sktime/sktime#7363]
**Is your feature request related to a problem? Please describe.**
So far, when detrending, we can use linear regression with polynomial features. I would like something more flexible, i.e. fiting linear splines in a way that Prophet does it, but without the probabilistic programming overhead and with more control about what happens when extrapolating.

**Describe the solution you'd like**
I would like a new class like the `PolynomialTrendForecaster`, but `SplineTrendForecaster` or so. Internally, it would look like the `PolynomialTrendForecaster`, with (hopefully) the only difference being that instead of
```python
self.regressor_ = make_pipeline(
    PolynomialFeatures(degree=self.degree, include_bias=self.with_intercept),
    regressor,
)
```

we would have something like

```python
self.regressor_ = make_pipeline(
    SplineTransformer(degree=self.degree, n_knots=self.knots, extrapolation=self.extrapolation, include_bias=self.with_intercept),
    regressor,
)
```

We could maybe even write another class `PipelineTrendForecaster` that takes a scikit-learn pipeline. From that, we could easily derive `PolynomialTrendForecaster` and `SplineTrendForecaster`.

scikit-learn's `SplineTransformer` has the advantage that you can tell it how to extrapolate, either linearly, constant, periodic, or more.

**Comment by fkiraly** [comment]
Great idea! I think this could be a good first issue.

Regarding the suggestion of `PipelineTrendForecaster`, this exists already if you are using the `Detrender` - take any time series transformation, and the pipeline `Detrender(trend_trafo) * my_forecaster` should give you what you want. This is syntactic sugar for a `TransformedTargetForecatser([Detrender(Trend_trafo), my_forecaster)`.

Thinking of which, should there be a `PolynomialTrend` transformer?

**Comment by fkiraly** [comment]
Just looked at `SplineFeatures` in `sklearn`, that is not a temporal spline smoother but a spline feature generator:
https://scikit-learn.org/1.5/modules/generated/sklearn.preprocessing.SplineTransformer.html

Maybe I am misunderstanding what it is what you would like?

The starting point is perhaps clarifying that/why the pipeline involving `PolynomialFeatures` concatenate with a forecaster is different from the `PolynomialTrendForecaster`.

**Comment by jgyasu** [comment]
Hi @fkiraly I want to work on this issue, I tried to understand the comments but I might need some help :)

**Comment by Garve** [comment]
Heya! Sorry for the late reply.

I mean the following:
If you use `SplineTransformer(degree=1, knots=3)` on the single column feature set `[0, 1, 2, 3, 4, ... t]`, you get two (n_knots-1) columns based on linear (since degree=1) B-splines.

You could then use linear regression to predict `y` using these two features. The result is a linear combination of the linear B-spline basis, which is a linear spline again.

**Comment by fkiraly** [comment]
@jgyasu, all yours! Let us know if you have any questions (I saw your comment on discord).

@Garve, I would still need a more detailed explanation on how we would use this for forecasting. Sorry if I am a bit slow here.

Could you produce a worked example for the following dataset?

index = 0, 1, 2, 3
values = 42, 43, 42, 43

We want to forecast two steps ahead, i.e., values at indices 4 and 5.

**Comment by Garve** [comment]
Hi! Sorry, finally got some time. I figured out that I can also use the TrendForecaster to get what I want.

```python
import pandas as pd
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import SplineTransformer
from sklearn.linear_model import LinearRegression

from sktime.datasets import load_airline
from sktime.forecasting.trend import TrendForecaster
from sktime.transformations.series.detrend import Detrender
from sktime.utils import plot_series

y = load_airline().iloc[:36]

spline_regressor = make_pipeline(SplineTransformer(degree=1, n_knots=4), LinearRegression())
spline_forecaster = TrendForecaster(spline_regressor)
spline_detrender = Detrender(spline_forecaster)

linear_detrender = Detrender()

plot_series(y, y - spline_detrender.fit_transform(y), y - linear_detrender.fit_transform(y), labels=["Data", "Spline Trend", "Linear Trend"])
```

Output:
![image](https://github.com/user-attachments/assets/4063b47f-0d25-49a3-8af6-a8d03c46f34d)

The spline trend is a linear combination of the following four B-splines:
```python
import numpy as np

spline_features = SplineTransformer(degree=1, n_knots=4).fit_transform(np.arange(36).reshape(-1, 1))
spline_data = pd.DataFrame(spline_features, index=y.index)
plot_series(*[spline_data[col] for col in spline_data.columns], labels=[f"Basis Spline #{i+1}" for i in range(len(spline_data.columns))])
```

![image](https://github.com/user-attachments/assets/162c2452-4e13-4402-b881-6026e1f4afe1)


This is kind of doing what the `ProphetPiecewiseLinearTrendForecaster` is doing, just without the MCMC overhead. Plus, you get also the flexibility to say what should happen extrapolating.

```python
from sktime.forecasting.naive import NaiveForecaster

y = load_airline().iloc[:36]

models = {}

for spline_extrapolation in ["linear", "constant", "periodic", "continue"]:
    spline_regressor = make_pipeline(SplineTransformer(degree=1, n_knots=4, extrapolation=spline_extrapolation), LinearRegression())
    spline_forecaster = TrendForecaster(spline_regressor)
    spline_detrender = Detrender(spline_forecaster)
    model = spline_detrender * NaiveForecaster()
    model.fit(y)
    models[spline_extrapolation] = model

plot_series(
    y,
    *[model.predict(fh=np.arange(1, 27)) for model in models.values()],
    labels=["Data", *[f"Prediction with {spline_extrapolation} trend extrapolation" for spline_extrapolation in models]]
)
```
![image](https://github.com/user-attachments/assets/15ce5444-67d8-48f1-b5e7-276793c1aeab)

Note that `"linear"` and `"constant`" result in the same extrapolation here since we used linear splines, but if we set `degree = 2`, we can see the following:


![image](https://github.com/user-attachments/assets/e57e8ccd-25db-4a43-ba41-2caa574df396)


The question is: do we want to create an extra `SplineTrendForecaster` for this, just like the `PolynomialTrendForecaster`?

**Comment by fkiraly** [comment]
Interesting! I get it now, you want to apply the splines to the index, so I also see now how the sklearn pipelining *inside* produces the required forecaster.

Then there is a question, @Garve - should we provide the user with an extra forecaster that is a shorthand for the pipeline, like `PolynomialTrendForecaster`, which exists as shorthand since it is popular?

Or, is it enough that we can user the more general `TrendForecaster` this way, perhaps add an example that explains the more complicated use case?

**Comment by Garve** [comment]
I think it's cool to have a shortcut for this one as well. It might not be as popular, but maybe it helps spreading this neat little idea. 😄

**Comment by fkiraly** [comment]
Great!

For @jgyasu, then what I would do is use the `_DelegatedForecaster` to create a shortcut for the above-mentioned forecaster, namely

```python
    spline_regressor = make_pipeline(SplineTransformer(**kwargs), regressor)
    spline_forecaster = TrendForecaster(spline_regressor)
```

with `kwargs` and `regressor` being parameters.

**Comment by Dehelaan** [comment]
@fkiraly can you assign this issue to me.

**Comment by fkiraly** [comment]
yes, this is a nice starter issue! Let us know if you need advice!

**Comment by Dehelaan** [comment]
@fkiraly in this issue, we just have to create a new trend forecaster from the scratch with base class as` _DelegatedForecaster`?

**Comment by fkiraly** [comment]
Yes, exactly - I would use the `_DelegatedForecaster`.

**Comment by jgyasu** [comment]
@fkiraly It's been two weeks since you assigned this issue to Dehelaan, I couldn't work on this because of my exams but I have been working on this now and facing some errors which I will fix shortly, can you please assign this to me?

**Comment by jgyasu** [comment]
@fkiraly I have already worked on this feature and it's been two weeks since this was assigned to someone else, if this is still open then I can open a PR
